File: images/ugc/src/api/v1/handlers/reviews.py
from http import HTTPStatus
from uuid import UUID

from fastapi import APIRouter, Depends, Path, Response

from api.dependencies import check_filmwork_exist, check_review_exists
from api.paginator import Paginator
from models.models import Review, ReviewRating, ReviewScore, Text
from services.auth import AuthService
from services.reviews import ReviewsService, get_reviews_service

router = APIRouter(tags=['reviews'])


@router.get(
    path='/filmworks/{filmwork_id}/reviews',
    summary='Просмотр списка рецензий',
    response_description='Список рецензий на фильм',
    response_model=list[Review],
    dependencies=[Depends(check_filmwork_exist)],
)
async def get_filmwork_reviews(
    service: ReviewsService = Depends(get_reviews_service),
    filmwork_id: UUID = Path(title='UUID фильма'),
    paginator: Paginator = Depends(),
) -> list[Review]:
    result = await service.filter(filmwork_id=filmwork_id, paginator=paginator)
    return result


@router.post(
    path='/filmworks/{filmwork_id}/reviews',
    summary='Добавление рецензии к фильму',
    response_description='Результат добавления рецензии на фильм',
    response_model=Review,
    response_model_exclude_none=True,
    dependencies=[Depends(check_filmwork_exist)],
)
async def create_filmwork_review(
    text: Text,
    auth: AuthService = Depends(),
    service: ReviewsService = Depends(get_reviews_service),
    filmwork_id: UUID = Path(title='UUID фильма'),
) -> Review:
    result = await service.update(user_id=auth.user_id, filmwork_id=filmwork_id, text=text.text)
    return result


@router.delete(
    path='/filmworks/{filmwork_id}/reviews/{review_id}',
    summary='Удаление рецензии у фильма',
    response_description='Результат удаления рецензии на фильму',
    dependencies=[Depends(check_filmwork_exist), Depends(check_review_exists)],
    status_code=HTTPStatus.NO_CONTENT,
)
async def delete_filmwork_review(
    auth: AuthService = Depends(),
    service: ReviewsService = Depends(get_reviews_service),
    filmwork_id: UUID = Path(title='UUID фильма'),
    review_id: UUID = Path(title='UUID рецензии'),
) -> Response:
    await service.remove(review_id=review_id, user_id=auth.user_id, filmwork_id=filmwork_id)
    return Response(status_code=HTTPStatus.NO_CONTENT)


@router.get(
    path='/filmworks/{filmwork_id}/reviews/{review_id}/ratings',
    summary='Просмотр рейтинга рецензии',
    response_description='Рейтинг рецензии (лайки, дизлайки и сумма)',
    response_model=ReviewRating,
    dependencies=[Depends(check_filmwork_exist)],
)
async def get_review_rating(
    service: ReviewsService = Depends(get_reviews_service),
    review_id: UUID = Path(title='UUID отзыва'),
    filmwork_id: UUID = Path(title='UUID фильма'),
) -> ReviewRating:
    result = await service.get_rating(filmwork_id=filmwork_id, review_id=review_id)
    return result


@router.post(
    path='/filmworks/{filmwork_id}/reviews/{review_id}/ratings',
    summary='Добавление оценки рецензии',
    response_description='Рейтинг рецензии (лайки, дизлайки и сумма)',
    response_model=ReviewRating,
    dependencies=[Depends(check_filmwork_exist)],
)
async def rate_review(
    score: ReviewScore,
    auth: AuthService = Depends(),
    service: ReviewsService = Depends(get_reviews_service),
    review_id: UUID = Path(title='UUID отзыва'),
    filmwork_id: UUID = Path(title='UUID фильма'),
) -> ReviewRating:
    result = await service.rate(
        filmwork_id=filmwork_id,
        review_id=review_id,
        user_id=auth.user_id,
        score=score.score,
    )
    return result


@router.delete(
    path='/filmworks/{filmwork_id}/reviews/{review_id}/ratings',
    summary='Удаление оценки у рецензии',
    response_model=ReviewRating,
    response_description='Результат удаления рейтинга рецензии',
    dependencies=[Depends(check_filmwork_exist)],
)
async def unrate_review(
    auth: AuthService = Depends(),
    service: ReviewsService = Depends(get_reviews_service),
    review_id: UUID = Path(title='UUID отзыва'),
    filmwork_id: UUID = Path(title='UUID фильма'),
) -> ReviewRating:
    result = await service.unrate(
        filmwork_id=filmwork_id,
        review_id=review_id,
        user_id=auth.user_id,
    )
    return result

File: images/ugc/internal/handlers/reviews.go
package handlers

import (
	"net/http"

	"github.com/google/uuid"
	"github.com/labstack/echo/v5"
	"ugc/internal/models"
	"ugc/internal/services"
)

type ReviewsHandler struct {
	reviewsService *services.ReviewsService
}

func NewReviewsHandler(reviewsService *services.ReviewsService) *ReviewsHandler {
	return &ReviewsHandler{reviewsService: reviewsService}
}

func (h *ReviewsHandler) GetFilmworkReviews(c *echo.Context) error {
	filmworkID, err := uuid.Parse(c.Param("filmwork_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid filmwork_id")
	}
	paginator := getPaginator(c)

	reviews, err := h.reviewsService.Filter(c.Request().Context(), filmworkID, paginator.Page, paginator.Size)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, reviews)
}

func (h *ReviewsHandler) CreateFilmworkReview(c *echo.Context) error {
	userID := c.Get("user_id").(uuid.UUID)
	filmworkID, err := uuid.Parse(c.Param("filmwork_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid filmwork_id")
	}

	var text models.Text
	if err := c.Bind(&text); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	review, err := h.reviewsService.Update(c.Request().Context(), userID, filmworkID, text.Text)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, review)
}

func (h *ReviewsHandler) DeleteFilmworkReview(c *echo.Context) error {
	userID := c.Get("user_id").(uuid.UUID)
	filmworkID, err := uuid.Parse(c.Param("filmwork_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid filmwork_id")
	}
	reviewID, err := uuid.Parse(c.Param("review_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid review_id")
	}

	err = h.reviewsService.Remove(c.Request().Context(), reviewID, userID, filmworkID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	return c.NoContent(http.StatusNoContent)
}

func (h *ReviewsHandler) GetReviewRating(c *echo.Context) error {
	filmworkID, err := uuid.Parse(c.Param("filmwork_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid filmwork_id")
	}
	reviewID, err := uuid.Parse(c.Param("review_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid review_id")
	}

	rating, err := h.reviewsService.GetRating(c.Request().Context(), reviewID, filmworkID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, rating)
}

func (h *ReviewsHandler) RateReview(c *echo.Context) error {
	userID := c.Get("user_id").(uuid.UUID)
	filmworkID, err := uuid.Parse(c.Param("filmwork_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid filmwork_id")
	}
	reviewID, err := uuid.Parse(c.Param("review_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid review_id")
	}

	var score models.ReviewScore
	if err := c.Bind(&score); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}
	if err := c.Validate(score); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	rating, err := h.reviewsService.Rate(c.Request().Context(), reviewID, userID, filmworkID, score.Score)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, rating)
}

func (h *ReviewsHandler) UnrateReview(c *echo.Context) error {
	userID := c.Get("user_id").(uuid.UUID)
	filmworkID, err := uuid.Parse(c.Param("filmwork_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid filmwork_id")
	}
	reviewID, err := uuid.Parse(c.Param("review_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid review_id")
	}

	rating, err := h.reviewsService.Unrate(c.Request().Context(), reviewID, filmworkID, userID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	return c.JSON(http.StatusOK, rating)
}
